src/test/java/test/io/ticktok/client/tick/RabbitTickerPolicyTest.java

```diff
@@ -0,0 +1,121 @@
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.ConnectionFactory;
+import io.ticktok.client.tick.rabbit.ChannelException;
+import io.ticktok.client.tick.TickChannel;
+import io.ticktok.client.tick.TickListener;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.function.Executable;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.ticktok.client.tick.rabbit.RabbitTickerPolicy.QUEUE_PARAM;
+import static io.ticktok.client.tick.rabbit.RabbitTickerPolicy.URI_PARAM;
+import static io.ticktok.client.tick.TickListener.RABBIT;
+import static java.lang.Thread.sleep;
+import static java.time.Duration.ofSeconds;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTimeoutPreemptively;
+
+class RabbitTickerPolicyTest {
+
+    private final TickListener tickListener = new TickListener();
+
+    @Test
+    void failWhenRabbitConnectionFailed() {
+        assertThrows(ChannelException.class, () -> {
+            listenOn("amqp://unknown");
+        });
+    }
+
+    private void listenOn(String uri) {
+        tickListener.forChannel(
+                TickChannel.builder()
+                        .type(RABBIT)
+                        .details(ImmutableMap.of(URI_PARAM, uri, QUEUE_PARAM, "kuku"))
+                        .build()).register(() -> {
+        });
+    }
+
+    @Test
+    void failOnInvalidUri() {
+        assertThrows(ChannelException.class, () -> {
+            listenOn("invalid uri");
+        });
+    }
+
+    @Test
+    void ignoreErrorWhenDisconnecting() {
+        tickListener.disconnect();
+    }
+
+    @Test
+    void disconnectAllClocks() throws Exception {
+        withQueues((channel, queueNames) -> {
+        AtomicInteger tickCount = new AtomicInteger();
+            queueNames.forEach(qn -> {
+                tickListener.forChannel(tickChannelFor(qn)).register(tickCount::incrementAndGet);
+            });
+
+            tickListener.disconnect();
+            for (String qn : queueNames) {
+                channel.basicPublish("", qn, null, "".getBytes());
+            }
+
+            sleep(1000);
+            assertThat(tickCount.get(), is(0));
+        }, "q1", "q2");
+    }
+
+    private TickChannel tickChannelFor(String qn) {
+        return TickChannel.builder()
+                .type(RABBIT)
+                .details(ImmutableMap.of(URI_PARAM, "amqp://localhost", QUEUE_PARAM, qn))
+                .build();
+    }
+
+    private void withQueues(WithQueuesCallable callable, String... queueNames) throws Exception {
+        Connection connection = new ConnectionFactory().newConnection();
+        Channel channel = connection.createChannel();
+        for (String qn : queueNames) {
```

Comment:
lets use more modern foreach...

Reply:
if I'll do that I would need to add an ugly try .. catch clause for it and I prefer not to

Reply:
am i missing something? i dont see any try catch here...

Reply:
am i missing something here? i dont see any try catch here...

Reply:
you don't that's why I stayed with the normal foreach so it can use the "throws Exception {"